### PR TITLE
Rewrite factorio server spawning and communication

### PR DESCRIPTION
**Pull request by Hornwitser**
This is a do over of the server spawning and communication logic, the idea is to get some structure on this instead of being a complete mess in client.js.  It's partly necessary for #200 as it proved to be problematic to adopt the new configuration system with all the clutter present.  It's also a hard requirement for #201.

I've not yet completed this, but I wanted to put this out in case anyone has some thoughts they would like to share on this.

### CODE REVIEW

File: client.js [hunk 1]
```diff
@@ -326,34 +229,84 @@ async function downloadMod() {
 	}
 }
 
+/**
```
Comment (psihius) [inline client.js]: Question, maybe this should go into the lib folder as utility methods? So we don't have big fat client.js with all the code and something more manageable?

File: client.js [hunk 2]
```diff
@@ -326,34 +229,84 @@ async function downloadMod() {
...
+				logger.warning(`Warning: ignoring file '${entry.name}' in sharedMods`);
+			}
+
+		} else {
+			logger.warning(`Warning: ignoring non-file '${entry.name}' in sharedMods`);
+		}
+	}
+}
+
 async function createInstance(config, args, instance) {
```
Comment (psihius) [inline client.js]: Same goes here. Also, this function is somewhat big, ideally you split it into smaller chunks and that also lends itself to go into a file of its own or something :)

File: client.js [hunk 3]
```diff
@@ -326,34 +229,84 @@ async function downloadMod() {
...
+	console.log(`Creating ${instance.path()}`);
+	await fs.ensureDir(instance.path());
+	await fs.ensureDir(instance.path("script-output"));
+
+	let target = path.join("lib", "scenarios");
+	let link = instance.path("scenarios");
+	await fs.symlink(path.relative(path.dirname(link), target), link, 'junction');
+
+	await symlinkMods(instance, "sharedMods", console);
 	let instconf = {
```
Comment (psihius) [inline client.js]: Maybe use actual files with default values, so they are not defined in the code like this and loaded via require?

File: client.js [hunk 4]
```diff
@@ -362,11 +315,16 @@ write-data=${ instance.path() }\r\n
...
+	await server.init();
+
+	let serverSettings = await server.exampleSettings();
 	let name = "Clusterio instance: " + instance.name;
 	if (config.username) {
 		name = config.username + "'s clusterio " + instance.name;
 	}
-	let serversettings = {
+
+	let overrides = {
```
Comment (psihius) [inline client.js]: Same here - load it from file?

File: client.js [hunk 5]
```diff
@@ -375,65 +333,55 @@ write-data=${ instance.path() }\r\n
...
-			process.exit(0);
+
+	for (let [name, value] of Object.entries(overrides)) {
+		if (!Object.hasOwnProperty.call(serverSettings, name)) {
+			throw Error(`Expected server settings to have a ${name} property`);
 		}
+		serverSettings[name] = value;
+	}
+
+	await fs.writeFile(instance.path("server-settings.json"), JSON.stringify(serverSettings, null, 4));
```
Comment (psihius) [inline client.js]: filename should be a `const server_config = "server-settings.json";`  or something like that. We have been working on removing hard-coded file names from the codebase.

File: client.js [hunk 6]
```diff
@@ -796,7 +680,7 @@ async function instanceManagement(config, instance, instanceconfig) {
 		setInterval(getID, 10000);
 		getID();
 		function getID() {
-			messageInterface("/silent-command rcon.print(#game.connected_players)", function(playerCount) {
+			server.sendRcon("/silent-command rcon.print(#game.connected_players)").then((playerCount) => {
```
Comment (psihius) [inline client.js]: But are we using only RCON going forward? stdin/stdout is a valid communication layer too and for some things probably is even better.

File: client.js (hunk 2 above)
Comment (Hornwitser) [reply client.js]: Unless you have some grand architectural structure pre planned, I feel like it's the wrong time to be moving these functions around.

**Comment by Hornwitser** [comment]
I've completed most of the code now.  Summary of changes:
- Refactored fileOps.getNewestFile as it was not compatible with the idea that you pass the name of the file as it appears in the saves directory.
- Renamed factorioPort to gamePort in the class.
- Removed some commented out code.
- Refactored getFactorioLocale and moved it to lib/factorio/locale
- Used the ini module to create the config.ini for Factorio
- Introduced a modular outputHeuristics filter which triggers actions based on outputs of Factorio.
- Changed the start and startScenario to wait until the server is ready to accept clients before returning.
- Added the disableAchievements method.
- Added integration tests.

**Comment by Hornwitser** [comment]
I've completed this now.  Latest changes are that I've removed the deriving of instance id from the RCON password, removed stringUtils, and fixed some issues with shutdown and rcon on Windows.

**Comment by Danielv123** [comment]
Your changelog is missing the fact that instanceconfig.unique is now deprecated in favour of instanceconfig.id - also, other places in the code its called instanceID, isn't it? Maybe consider changing that if it makes sense

**Comment by Hornwitser** [comment]
I'm pretty sure the config rewrite is going to have change this again so keeping it as unique should be fine for now.  I didn't want to name it unique in the config file and calling it instanceconfig.instanceID seems a little redundant.

File: client.js (hunk 5 above)
Comment (psihius) [reply client.js]: The thing is - you define the file names somewhere at the top of the file, preferably the root file and you just use that variable throughout the code. And it makes it easier to also add the ability to pass the file name as a command-line parameter. While we use instance-per-dir approach, you might want to use more than one server-settings.json file or even store them somewhere else and pass them at launch as paths.

**Pull request merged**